### DISCUSSION TITLE
Realign expressions and assertion methods

### DISCUSSION THREAD

**Discussion by BernieWhite** [microsoft/PSRule#904] [Ideas]
Across expressions (used in YAML and JSON resources) and assertion methods (used in PowerShell resources) there are different names used to equivalent functionality. While terminology differs between programming/ scripting languages we want to make it even easier to understand and author rules.

We should look to realign naming to be ideally the same for expressions and assertion methods.

This discussion proposes the following naming changes.

Expression      | Assertion method | Proposed change
----------      | ---------------- | ------------------
Contains        | Contains         | _No change_
Count           | Count            | _No change_
Equals          | _n/a_   | See note _(1)_
EndsWith        | EndsWith         | _No change_
Exists          | HasField         | Rename assertion method to `Exists`.
Greater         | Greater          | _No change_
GreaterOrEquals | GreaterOrEqual   | Rename expression to `GreaterOrEqual`.
HasDefault      | HasDefaultValue  | Rename assertion method to `HasDefault`.
HasSchema       | HasJsonSchema    | Rename assertion method to `HasSchema`.
HasValue        | _n/a_            | See note _(1)_.
In              | In               | _No change_
IsLower         | IsLower          | _No change_
IsString        | IsString         | _No change_
IsUpper         | IsUpper          | _No change_
Less            | Less             | _No change_
LessOrEquals    | LessOrEqual      | Rename expression to `LessOrEqual`.
Match           | Match            | _No change_
NotEquals       | _n/a_            | _No change_
NotIn           | NotIn            | _No change_
NotMatch        | NotMatch         | _No change_
SetOf           | SetOf            | _No change_
StartsWith      | StartsWith       | _No change_
Subset          | Subset           | _No change_
Version         | Version          | _No change_
_n/a_           | FileHeader       | _No change_
_n/a_           | FilePath         | _No change_
_n/a_           | HasFields        | _No change_
_n/a_           | HasFieldValue    | See note _(1)_
IsArray        | IsArray          | See note _(2)_
IsBoolean    | IsBoolean        | See note _(2)_
IsDateTime | IsDateTime       | See note _(2)_
IsInteger     | IsInteger        | See note _(2)_
IsNumeric   | IsNumeric        | See note _(2)_
_n/a_           | JsonSchema       | _No change_
_n/a_           | NotHasField      | _No change_
_n/a_           | NotNull          | _No change_
NotWithinPath | NotWithinPath    | See note _(3)_
_n/a_           | Null             | _No change_
_n/a_           | NullOrEmpty      | _No change_
_n/a_           | TypeOf           | _No change_
WithinPath | WithinPath       | See note _(3)_

- (1) The `Equals`, `HasValue` and expression and `HasFieldValue` are approximate. Also these are not necessarily intuitive. 
  - It may be better to introduce `Empty`, `NotEmpty`, and `Equal` across expressions and assertion methods. Then deprecate `Equals`, `HasValue` and `HasFieldValue`.
- (2) Added in #908.
- (3) Added in #933.

https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Expressions/#hasvalue
https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Expressions/#equals
https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Assert/#hasfieldvalue

**Comment by BernieWhite** [comment]
@ArmaanMcleod Thoughts?

**Reply by ArmaanMcleod** [reply]
@BernieWhite Looks good. I think there is benefit to bringing in those new assertions for (1). `Empty` and `NotEmpty` would make it more intuitive to check empty values.
